fix(ios): format js errors in cli output

janvennemann opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-27812

This properly formats the error output in the CLI and prints JS/native stack traces from any runtime errors. The format is similar to Node/V8 with added native stack trace. It will also prevent showing the unrelated error modal for "Module xy failed to leave a valid exports object".

janvennemann committed:
fix(ios): format js errors in cli output

sgtcoolguy commented:
@garymathews should weigh in here since this likely affects ACA/crash reporting. It looks to me like the API/header change may be a breaking change as well.

build commented:
<!--
  3 failure:  :microscope: Ther..., Test suite crashe..., Tests have failed...
  1 warning:  This PR has miles...
  2 messages
  1 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/janvennemann/titanium_mobile/blob/TIMOB-27812/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Test suite crashed on iOS simulator. Please see <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-12139/1/artifact/mocha_2020-09-28-125039_macos-walle.crash">the crash log</a> for more details.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This PR has milestone set to 9.2.1, but the version defined in package.json is 9.3.0
Please either:
- Update the milestone on the PR
- Update the version in package.json
- Hold the PR to be merged later after a release and version bump on this branch</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 9 tests have failed
There are 9 tests failing and 742 skipped out of 8167 total tests.</td>
    </tr>
  </tbody>
</table>

### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>ios.macos.Titanium.Blob</td><td>image dimensions should be reported in pixels (10.15.5)</td><td>0.002</td><td><pre>Error: expected 6 to be 11
fail@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:356:23
postlayout@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/ti.blob.test.js:483:33</pre></td></tr>
<tr><td>ios.macos.Titanium.UI.Semantic Colors</td><td>use semantic colors via color properties (10.15.5)</td><td>0.042</td><td><pre>Error: expected 'Ti.UI.View' view to match snapshot image: snapshots/systemredcolor_light.png
    width
fail@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:356:23
postlayout@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/ti.ui.test.js:363:34</pre></td></tr>
<tr><td>ios.macos.Titanium.UI.View.borderRadius corners</td><td>1 value with shadow effect (10.15.5)</td><td>0.005</td><td><pre>Error: expected 'Ti.UI.View' view to match snapshot image: snapshots/borderRadiusWithShadow30px_1x.png
    width
fail@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:356:23
postlayout@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/ti.ui.view.test.js:1232:39</pre></td></tr>
<tr><td>ios.macos.Titanium.UI.View.borderRadius corners</td><td>1 value to create circle (10.15.5)</td><td>0.004</td><td><pre>Error: expected 'Ti.UI.View' view to match snapshot image: snapshots/borderRadius30px.png
    width
fail@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:356:23
postlayout@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/ti.ui.view.test.js:1200:39</pre></td></tr>
<tr><td>ios.macos.Titanium.UI.View.borderRadius corners</td><td>set property post layout (10.15.5)</td><td>0.016</td><td><pre>Error: expected 'Ti.UI.View' view to match snapshot image: snapshots/borderRadius12px_12_1x.png
    width
fail@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:356:23
postlayout@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/ti.ui.view.test.js:1169:39</pre></td></tr>
<tr><td>ios.macos.Titanium.UI.View.borderRadius corners</td><td>2 values in Array (10.15.5)</td><td>0.004</td><td><pre>Error: expected 'Ti.UI.View' view to match snapshot image: snapshots/borderRadius12px_12_1x.png
    width
fail@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:356:23
postlayout@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/ti.ui.view.test.js:1135:41</pre></td></tr>
<tr><td>ios.macos.Titanium.UI.View.borderRadius corners</td><td>2 values in String (10.15.5)</td><td>0.004</td><td><pre>Error: expected 'Ti.UI.View' view to match snapshot image: snapshots/borderRadius12px_12_1x.png
    width
fail@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:356:23
postlayout@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/ti.ui.view.test.js:1101:41</pre></td></tr>
<tr><td>ios.macos.Titanium.UI.View.borderRadius corners</td><td>4 values in Array (10.15.5)</td><td>0.005</td><td><pre>Error: expected 'Ti.UI.View' view to match snapshot image: snapshots/borderRadius12px_12_12dp_12_1x.png
    width
fail@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:356:23
postlayout@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/ti.ui.view.test.js:1068:41</pre></td></tr>
<tr><td>ios.macos.Titanium.UI.View.borderRadius corners</td><td>4 values in String (10.15.5)</td><td>0.007</td><td><pre>Error: expected 'Ti.UI.View' view to match snapshot image: snapshots/borderRadius12px_12_12dp_12_1x.png
    width
fail@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/node_modules/should/cjs/should.js:356:23
postlayout@file:///Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-12139/tmp/mocha/build/iphone/build/Products/Debug-maccatalyst/mocha.app/Contents/Resources/ti.ui.view.test.js:1034:41</pre></td></tr>
</table>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 240acb57282e5f5af433c124abf143509030b6e9
</p>

janvennemann commented:
@sgtcoolguy @garymathews i reverted the public API change and just replaced the formatting inside the `description` method.

vijaysingh-axway commented on iphone/TitaniumKit/TitaniumKit/Sources/API/TiExceptionHandler.m:
```diff
@@ -94,45 +94,15 @@ - (void)reportScriptError:(JSValue *)error inJSContext:(JSContext *)context
-
-    [[TiApp app] showModalError:[NSString stringWithFormat:@"%@\n\n%@", [error description], stackTrace]];
-  }
+  NSString *stackTrace = [error.formattedNativeStack componentsJoinedByString:@"\n"];
```
`formattedNativeStack` is declared as a function and it is getting used as property . Though it can be used, I think it's better to call it as function to have more readability. 
```suggestion
  NSString *stackTrace = [[error formattedNativeStack] componentsJoinedByString:@"\n"];
```

vijaysingh-axway commented on iphone/TitaniumKit/TitaniumKit/Sources/API/TiExceptionHandler.m:
```diff
@@ -197,40 +174,111 @@ - (void)dealloc
-    // remove bundle path from source paths
-    NSString *encodedBundlePath = [NSString stringWithFormat:@"file://%@", [[NSBundle mainBundle].bundlePath stringByReplacingOccurrencesOfString:@" " withString:@"%20"]];
-    NSString *jsStack = [self.backtrace stringByReplacingOccurrencesOfString:encodedBundlePath withString:@""];
+  [message appendFormat:@"\n\nNative Stack:\n    %@", [self.formattedNativeStack componentsJoinedByString:@"\n    "]];
```
Sam as above.
```suggestion
  [message appendFormat:@"\n\nNative Stack:\n    %@", [[self formattedNativeStack] componentsJoinedByString:@"\n    "]];
```

vijaysingh-axway committed:
Merge branch 'master' into TIMOB-27812

vijaysingh-axway reviewed (commented):
There is a minor suggestion.  Otherwise it looks good to me.
@garymathews Can you look in this PR as it is related to ACA?

garymathews reviewed (commented):
Comparing Android to iOS, here's the error outputs:
##### iOS
```
[ERROR] /alloy/controllers/index.js:129
[ERROR]       Titanium.crash();
[ERROR]                      ^
[ERROR]
[ERROR] TypeError: Titanium.crash is not a function. (In 'Titanium.crash()', 'Titanium.crash' is undefined)
[ERROR]     at test (/alloy/controllers/index.js:129:21)
[ERROR]
[ERROR] Native Stack:
[ERROR]     TitaniumKit 0x0000000100dffb77 -[TiScriptError description] + 1127
[ERROR]     Foundation 0x00007fff20862a33 _NSDescriptionWithLocaleFunc + 82
[ERROR]     CoreFoundation 0x00007fff203d34cd __CFStringAppendFormatCore + 12102
[ERROR]     CoreFoundation 0x00007fff203d4f61 _CFStringCreateWithFormatAndArgumentsReturningMetadata + 145
[ERROR]     CoreFoundation 0x00007fff203d4eca _CFStringCreateWithFormatAndArgumentsAux2 + 20
[ERROR]     TitaniumKit 0x0000000100dbf7a8 TiLogMessage + 175
[ERROR]     TitaniumKit 0x0000000100dff12c -[TiExceptionHandler reportScriptError:] + 74
[ERROR]     TitaniumKit 0x0000000100debe06 TiBindingEventProcess + 731
[ERROR]     TitaniumKit 0x0000000100e0d4eb -[KrollContext invoke:] + 109
[ERROR]     TitaniumKit 0x0000000100e0d55e -[KrollContext enqueue:] + 93
[ERROR]     TitaniumKit 0x0000000100e14a6e TiBindingRunLoopEnqueue + 77
[ERROR]     TitaniumKit 0x0000000100e132cd -[TiProxy fireEvent:withObject:propagate:reportSuccess:errorCode:message:] + 164
[ERROR]     TitaniumKit 0x0000000100df3b5f -[TiViewProxy fireEvent:withObject:propagate:reportSuccess:errorCode:message:] + 224
[ERROR]     ACA Tester 0x00000001009397b0 -[TiUIButtonProxy fireEvent:withObject:propagate:reportSuccess:errorCode:message:] + 272
[ERROR]     TitaniumKit 0x0000000100e12ff8 -[TiProxy fireEvent:withObject:] + 72
[ERROR]     ACA Tester 0x000000010093c2de -[TiUIButton controlAction:forEvent:] + 686
[ERROR]     UIKitCore 0x00007fff246670af -[UIApplication sendAction:to:from:forEvent:] + 83
[ERROR]     UIKitCore 0x00007fff23f94584 -[UIControl sendAction:to:forEvent:] + 223
[ERROR]     UIKitCore 0x00007fff23f948a7 -[UIControl _sendActionsForEvents:withEvent:] + 332
```

##### Android
```
[ERROR] TiExceptionHandler: (main) [2,8113] /alloy/controllers/index.js:129
[ERROR] TiExceptionHandler:       Titanium.crash();
[ERROR] TiExceptionHandler:                ^
[ERROR] TiExceptionHandler: TypeError: Titanium.crash is not a function
[ERROR] TiExceptionHandler:     at Button.test (/alloy/controllers/index.js:129:16)
[ERROR] TiExceptionHandler:     at Button.value (ti:/events.js:50:21)
[ERROR] TiExceptionHandler:     at Button.value (ti:/events.js:102:19)
[ERROR] TiExceptionHandler:
[ERROR] TiExceptionHandler:     org.appcelerator.kroll.runtime.v8.V8Object.nativeFireEvent(Native Method)
[ERROR] TiExceptionHandler:     org.appcelerator.kroll.runtime.v8.V8Object.fireEvent(V8Object.java:63)
[ERROR] TiExceptionHandler:     org.appcelerator.kroll.KrollProxy.doFireEvent(KrollProxy.java:977)
[ERROR] TiExceptionHandler:     org.appcelerator.kroll.KrollProxy.handleMessage(KrollProxy.java:1206)
[ERROR] TiExceptionHandler:     org.appcelerator.titanium.proxy.TiViewProxy.handleMessage(TiViewProxy.java:264)
[ERROR] TiExceptionHandler:     android.os.Handler.dispatchMessage(Handler.java:102)
[ERROR] TiExceptionHandler:     android.os.Looper.loop(Looper.java:223)
[ERROR] TiExceptionHandler:     android.app.ActivityThread.main(ActivityThread.java:7523)
[ERROR] TiExceptionHandler:     java.lang.reflect.Method.invoke(Native Method)
[ERROR] TiExceptionHandler:     com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
```

There's an additional `\n` under the line arrow pointing to the error, and Android does not contain a `Native Stack:` header. I'd prefer to have iOS match Android.

janvennemann committed:
fix: address review feedback

janvennemann commented:
@garymathews i changed the formatting of the error message to match Android
@vijaysingh-axway Instead of calling it as a method i converted it to an actual readonly property with custom getter, so things should be clear now

vijaysingh-axway reviewed (approved):
CR passed.

ssekhri commented:
FR Passed
Mac OS: 10.15.4
SDK: 9.3.0.v20200930062956
Appc CLI: 8.1.1
JDK: 11.0.4
Node: 10.17.0
Xcode: 12.0
